File: CSMApps/Vertebrae/v1.0.0/src/main/java/com/carrpod/vertebrae/comm/SessionCommunicator.java
package com.carrpod.vertebrae.comm;

import android.content.Context;
import android.net.LocalSocket;
import android.net.LocalSocketAddress;
import android.util.Log;

import com.carrpod.vertebrae.model.InterSessionMessage;
import com.carrpod.vertebrae.model.KilosSession;
import com.carrpod.vertebrae.storage.SessionStorageManager;
import com.google.gson.Gson;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class SessionCommunicator {

    private static final String TAG = "SessionCommunicator";
    private static SessionCommunicator instance;

    private final Context context;
    private final SessionStorageManager storage;
    private final Gson gson = new Gson();
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final String localSessionId = "vertebrae_" + System.currentTimeMillis();
    private LocalSocket socket;
    private BufferedReader reader;
    private BufferedWriter writer;
    private boolean isConnected = false;
    private final List<MessageListener> listeners = new CopyOnWriteArrayList<>();

    public interface MessageListener {
        void onMessage(InterSessionMessage message);
        void onConnected();
        void onDisconnected();
        void onError(String error);
    }

    private SessionCommunicator(Context context) {
        this.context = context.getApplicationContext();
        this.storage = SessionStorageManager.getInstance();
    }

    public static synchronized void initialize(Context context) {
        if (instance == null) {
            instance = new SessionCommunicator(context);
        }
    }

    public static SessionCommunicator getInstance() {
        if (instance == null) {
            throw new IllegalStateException("SessionCommunicator not initialized");
        }
        return instance;
    }

    public void addListener(MessageListener listener) {
        listeners.add(listener);
    }

    public void removeListener(MessageListener listener) {
        listeners.remove(listener);
    }

    public void connect() {
        if (isConnected) return;

        executor.execute(() -> {
            try {
                socket = new LocalSocket();
                socket.connect(new LocalSocketAddress(InterSessionServer.SOCKET_NAME, LocalSocketAddress.Namespace.RESERVED));

                reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
                writer = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream()));

                // Register this client
                writer.write("REGISTER:" + localSessionId + "\n");
                writer.flush();

                isConnected = true;
                notifyConnected();

                // Read loop
                String line;
                while (isConnected && (line = reader.readLine()) != null) {
                    if (line.isEmpty()) continue;

                    try {
                        InterSessionMessage message = gson.fromJson(line, InterSessionMessage.class);
                        if (message != null && message.isForSession(localSessionId)) {
                            notifyMessage(message);
                        }
                    } catch (Exception e) {
                        Log.e(TAG, "Parse error", e);
                    }
                }
            } catch (Exception e) {
                Log.e(TAG, "Connection error", e);
                notifyError(e.getMessage() != null ? e.getMessage() : "Connection failed");
            } finally {
                disconnect();
            }
        });
    }

    public void disconnect() {
        isConnected = false;
        try { if (reader != null) reader.close(); } catch (Exception ignored) {}
        try { if (writer != null) writer.close(); } catch (Exception ignored) {}
        try { if (socket != null) socket.close(); } catch (Exception ignored) {}
        notifyDisconnected();
    }

    public void send(InterSessionMessage message) {
        if (!isConnected || writer == null) return;

        executor.execute(() -> {
            try {
                writer.write(gson.toJson(message));
                writer.newLine();
                writer.flush();
            } catch (Exception e) {
                Log.e(TAG, "Send failed", e);
            }
        });
    }

    public void sendText(String toSessionId, String groupId, String text) {
        InterSessionMessage msg = new InterSessionMessage();
        msg.setFromSessionId(localSessionId);
        msg.setToSessionId(toSessionId);
        msg.setGroupId(groupId);
        msg.setType(InterSessionMessage.MessageType.TEXT);
        msg.setPayload(text);
        send(msg);
    }

    public void sendCommand(String toSessionId, String groupId, String command, java.util.Map<String, String> params) {
        InterSessionMessage msg = new InterSessionMessage();
        msg.setFromSessionId(localSessionId);
        msg.setToSessionId(toSessionId);
        msg.setGroupId(groupId);
        msg.setType(InterSessionMessage.MessageType.COMMAND);
        msg.setPayload(gson.toJson(new CommandPayload(command, params)));
        send(msg);
    }

    public void sendFile(String groupId, java.io.File file, String toSessionId) {
        executor.execute(() -> {
            try {
                java.security.MessageDigest digest = java.security.MessageDigest.getInstance("SHA-256");
                byte[] hashBytes = digest.digest(java.nio.file.Files.readAllBytes(file.toPath()));
                StringBuilder hash = new StringBuilder();
                for (byte b : hashBytes) {
                    hash.append(String.format("%02x", b));
                }

                InterSessionMessage.FileAttachment attachment = new InterSessionMessage.FileAttachment(
                        file.getName(), file.getAbsolutePath(), file.length(),
                        context.getContentResolver().getType(android.net.Uri.fromFile(file)),
                        hash.toString()
                );

                // Save to shared storage
                storage.saveReceivedFile(groupId, file.getName(), new java.io.FileInputStream(file));

                InterSessionMessage msg = new InterSessionMessage();
                msg.setFromSessionId(localSessionId);
                msg.setToSessionId(toSessionId);
                msg.setGroupId(groupId);
                msg.setType(InterSessionMessage.MessageType.FILE_TRANSFER);
                msg.setPayload(gson.toJson(attachment));
                msg.setFileAttachment(attachment);
                send(msg);
            } catch (Exception e) {
                Log.e(TAG, "File send failed", e);
            }
        });
    }

    public void requestFile(String groupId, String fileName, String fromSessionId) {
        InterSessionMessage msg = new InterSessionMessage();
        msg.setFromSessionId(localSessionId);
        msg.setToSessionId(fromSessionId);
        msg.setGroupId(groupId);
        msg.setType(InterSessionMessage.MessageType.FILE_REQUEST);
        msg.setPayload(fileName);
        send(msg);
    }

    public void sendHeartbeat(String groupId, KilosSession session) {
        InterSessionMessage msg = new InterSessionMessage();
        msg.setFromSessionId(localSessionId);
        msg.setToSessionId(null);
        msg.setGroupId(groupId);
        msg.setType(InterSessionMessage.MessageType.HEARTBEAT);
        msg.setPayload(gson.toJson(new SessionHeartbeatPayload(session.getId(), session.getStatus(), session.isFocused())));
        send(msg);
    }

    public void sendFocusChange(String groupId, String sessionId, boolean isFocused) {
        InterSessionMessage msg = new InterSessionMessage();
        msg.setFromSessionId(sessionId);
        msg.setToSessionId(null);
        msg.setGroupId(groupId);
        msg.setType(InterSessionMessage.MessageType.FOCUS_CHANGE);
        msg.setPayload(gson.toJson(new FocusPayload(isFocused)));
        send(msg);
    }

    public void sendWindowState(String groupId, String sessionId, com.carrpod.vertebrae.model.WindowState state) {
        InterSessionMessage msg = new InterSessionMessage();
        msg.setFromSessionId(sessionId);
        msg.setToSessionId(null);
        msg.setGroupId(groupId);
        msg.setType(InterSessionMessage.MessageType.WINDOW_STATE);
        msg.setPayload(gson.toJson(state));
        send(msg);
    }

    private void notifyConnected() {
        for (MessageListener listener : listeners) {
            try { listener.onConnected(); } catch (Exception e) {}
        }
    }

    private void notifyDisconnected() {
        for (MessageListener listener : listeners) {
            try { listener.onDisconnected(); } catch (Exception e) {}
        }
    }

    private void notifyMessage(InterSessionMessage message) {
        for (MessageListener listener : listeners) {
            try { listener.onMessage(message); } catch (Exception e) {}
        }
    }

    private void notifyError(String error) {
        for (MessageListener listener : listeners) {
            try { listener.onError(error); } catch (Exception e) {}
        }
    }

    public void destroy() {
        disconnect();
        executor.shutdown();
    }

    // Payload classes
    private static class CommandPayload {
        public String command;
        public java.util.Map<String, String> params;

        CommandPayload(String command, java.util.Map<String, String> params) {
            this.command = command;
            this.params = params;
        }
    }

    private static class SessionHeartbeatPayload {
        public String sessionId;
        public KilosSession.SessionStatus status;
        public boolean isFocused;

        SessionHeartbeatPayload(String sessionId, KilosSession.SessionStatus status, boolean isFocused) {
            this.sessionId = sessionId;
            this.status = status;
            this.isFocused = isFocused;
        }
    }

    private static class FocusPayload {
        public boolean focused;

        FocusPayload(boolean focused) {
            this.focused = focused;
        }
    }
}

File: CSMApps/Vertebrae/v1.0.0/src/main/kotlin_backup/SessionCommunicator.kt
package com.carrpod.vertebrae.comm

import android.content.Context
import android.net.LocalSocket
import android.net.LocalSocketAddress
import android.util.Log
import com.carrpod.vertebrae.model.InterSessionMessage
import com.carrpod.vertebrae.model.KilosSession
import com.carrpod.vertebrae.storage.SessionStorageManager
import kotlinx.coroutines.*
import java.io.*

class SessionCommunicator private constructor(private val context: Context) {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val storage = SessionStorageManager.getInstance()
    private var socket: LocalSocket? = null
    private var output: BufferedWriter? = null
    private var input: BufferedReader? = null
    private val sessionId = "vertebrae_${System.currentTimeMillis()}"
    private val listeners = mutableListOf<MessageListener>()
    private var isConnected = false

    interface MessageListener {
        fun onMessage(message: InterSessionMessage)
        fun onConnected()
        fun onDisconnected()
        fun onError(error: String)
    }

    fun addListener(listener: MessageListener) {
        listeners.add(listener)
    }

    fun removeListener(listener: MessageListener) {
        listeners.remove(listener)
    }

    fun connect() {
        if (isConnected) return
        scope.launch {
            try {
                socket = LocalSocket()
                socket?.connect(LocalSocketAddress(InterSessionServer.SOCKET_NAME, LocalSocketAddress.Namespace.RESERVED))
                output = BufferedWriter(OutputStreamWriter(socket!!.outputStream))
                input = BufferedReader(InputStreamReader(socket!!.inputStream))

                // Register this session
                output!!.write("REGISTER:$sessionId\n")
                output!!.flush()

                isConnected = true
                scope.launch(Dispatchers.Main) { listeners.forEach { it.onConnected() } }

                // Read loop
                while (isConnected) {
                    val line = input?.readLine() ?: break
                    if (line.isNotEmpty()) {
                        val msg = com.google.gson.Gson().fromJson(line, InterSessionMessage::class.java)
                        msg?.let { scope.launch(Dispatchers.Main) { listeners.forEach { it.onMessage(it) } } }
                    }
                }
            } catch (e: Exception) {
                Log.e("SessionComm", "Connection error", e)
                scope.launch(Dispatchers.Main) { listeners.forEach { it.onError(e.message ?: "Connection error") } }
            } finally {
                disconnect()
            }
        }
    }

    fun send(message: InterSessionMessage) {
        if (!isConnected) return
        scope.launch(Dispatchers.IO) {
            try {
                output?.write(com.google.gson.Gson().toJson(message))
                output?.newLine()
                output?.flush()
            } catch (e: Exception) {
                Log.e("SessionComm", "Send failed", e)
            }
        }
    }

    fun sendText(toSessionId: String?, groupId: String, text: String) {
        val msg = InterSessionMessage(
            fromSessionId = sessionId,
            toSessionId = toSessionId,
            groupId = groupId,
            type = InterSessionMessage.MessageType.TEXT,
            payload = text,
        )
        send(msg)
    }

    fun sendCommand(toSessionId: String?, groupId: String, command: String, params: Map<String, String>) {
        val payload = com.google.gson.Gson().toJson(mapOf("command" to command, "params" to params))
        val msg = InterSessionMessage(
            fromSessionId = sessionId,
            toSessionId = toSessionId,
            groupId = groupId,
            type = InterSessionMessage.MessageType.COMMAND,
            payload = payload,
        )
        send(msg)
    }

    fun sendFile(groupId: String, file: java.io.File, toSessionId: String? = null) {
        scope.launch(Dispatchers.IO) {
            try {
                val hash = file.digest("SHA-256").joinToString("") { "%02x".format(it) }
                val attachment = InterSessionMessage.FileAttachment(
                    name = file.name,
                    path = file.absolutePath,
                    size = file.length(),
                    mimeType = context.contentResolver.getType(android.net.Uri.fromFile(file)) ?: "application/octet-stream",
                    hash = hash,
                )
                // Copy to shared storage
                val stored = storage.saveReceivedFile(groupId, file.name, file.inputStream()).await()

                val msg = InterSessionMessage(
                    fromSessionId = sessionId,
                    toSessionId = toSessionId,
                    groupId = groupId,
                    type = InterSessionMessage.MessageType.FILE_TRANSFER,
                    payload = com.google.gson.Gson().toJson(attachment),
                    fileAttachment = attachment,
                )
                send(msg)
            } catch (e: Exception) {
                Log.e("SessionComm", "File send failed", e)
            }
        }
    }

    fun requestFile(groupId: String, fileName: String, fromSessionId: String?) {
        val msg = InterSessionMessage(
            fromSessionId = sessionId,
            toSessionId = fromSessionId,
            groupId = groupId,
            type = InterSessionMessage.MessageType.FILE_REQUEST,
            payload = fileName,
        )
        send(msg)
    }

    fun sendHeartbeat(groupId: String, session: KilosSession) {
        val msg = InterSessionMessage(
            fromSessionId = sessionId,
            toSessionId = null,
            groupId = groupId,
            type = InterSessionMessage.MessageType.HEARTBEAT,
            payload = com.google.gson.Gson().toJson(SessionHeartbeat(
                sessionId = session.id,
                status = session.status,
                isFocused = session.isFocused,
            )),
        )
        send(msg)
    }

    fun sendFocusChange(groupId: String, sessionId: String, isFocused: Boolean) {
        val msg = InterSessionMessage(
            fromSessionId = sessionId,
            toSessionId = null,
            groupId = groupId,
            type = InterSessionMessage.MessageType.FOCUS_CHANGE,
            payload = com.google.gson.Gson().toJson(mapOf("focused" to isFocused)),
        )
        send(msg)
    }

    fun sendWindowState(groupId: String, sessionId: String, state: WindowState) {
        val msg = InterSessionMessage(
            fromSessionId = sessionId,
            toSessionId = null,
            groupId = groupId,
            type = InterSessionMessage.MessageType.WINDOW_STATE,
            payload = com.google.gson.Gson().toJson(state),
        )
        send(msg)
    }

    fun disconnect() {
        isConnected = false
        try { input?.close() } catch (e: Exception) {}
        try { output?.close() } catch (e: Exception) {}
        try { socket?.close() } catch (e: Exception) {}
        scope.launch(Dispatchers.Main) { listeners.forEach { it.onDisconnected() } }
    }

    fun destroy() {
        disconnect()
        scope.cancel()
    }

    companion object {
        private var INSTANCE: SessionCommunicator? = null
        fun initialize(context: Context) {
            INSTANCE = SessionCommunicator(context.applicationContext)
        }
        fun getInstance(): SessionCommunicator = INSTANCE!!
    }
}